### PR TITLE
[inductor] Match commuted attention-mask addition in masked SDPA fusion

### LINKED ISSUE #195782: [TorchInductor] Masked SDPA fusion misses commuted mask addition
### 🐛 Describe the bug

Masked SDPA fusion patterns encode the attention-mask addition as `scores + attn_mask` and miss the equivalent `attn_mask + scores` form. I reproduced this with `_sfdp_pattern_25` and `_sfdp_pattern_14`. The baseline is fused, while exchanging only the two operands leaves the attention computation unfused.

## Minimal reproducer

The reproducer is based on the upstream `_test_sdpa_rewriter_25` case. It uses FP16 inputs with shape `(4, 2, 16, 32)`, an attention mask with shape `(1, 1, 1, 2)`, and the same softmax and inference-dropout structure.

```python
"""Positive case: score + attn_mask."""
from __future__ import annotations

import torch
import torch.nn.functional as F
from torch._dynamo.utils import counters


class Model(torch.nn.Module):
    def forward(self, q, k, v, mask, training):
        q = q.transpose(1, 2)
        k = k.transpose(1, 2)
        v = v.transpose(1, 2)
        scores = q @ k.permute(0, 1, 3, 2)
        masked_score = scores + mask
        return F.dropout(
            masked_score.float().softmax(-1).type(v.dtype), p=0.1, training=training
        ) @ v


def main():
    torch.manual_seed(1845)
    device = torch.device("cpu")
    model = Model().to(device).eval()
    qkv = tuple(
        torch.randn(4, 2, 16, 32, dtype=torch.half, device=device)
        for _ in range(3)
    )
    inputs = (*qkv, torch.randn(1, 1, 1, 2, dtype=torch.half, device=device))
    compiled_inputs = [x.detach().clone() for x in inputs]
    training = False
    torch.manual_seed(1846)
    eager = model(*inputs, training)

    torch._dynamo.reset()
    counters.clear()
    torch.manual_seed(1846)
    compiled = torch.compile(model, backend="inductor", fullgraph=True)(
        *compiled_inputs, training
    )
    torch.testing.assert_close(compiled, eager, atol=2e-3, rtol=2e-3)
    print("fuse_attention:", counters["inductor"]["fuse_attention"])


if __name__ == "__main__":
    main()
```

The commuted-operand case is identical except for the following single-line change:

```diff
- masked_score = scores + mask
+ masked_score = mask + scores
```

## Observed behavior

**Instrumentation**

The baseline matches `_sfdp_pattern_25`, while the commuted form does not.

```text
scores + mask: HIT _sfdp_pattern_25 
mask + scores: no HIT _sfdp_pattern_25
```

**Graph**

The post-joint graphs confirm the instrumentation result:

```text
scores + mask:
  aten._scaled_dot_product_flash_attention_for_cpu.default

mask + scores:
  bmm → add → softmax decomposition → bmm
```

## Independent verification with pattern 14

I also repeated the same phenomenon for `_sfdp_pattern_14`.

```python
"""Positive control for _sfdp_pattern_14: scores + attn_mask."""
from __future__ import annotations

import torch
from torch._dynamo.utils import counters


class Model(torch.nn.Module):
    def forward(self, query, key, value):
        attn_mask = torch.ones(
            query.size(1), key.size(1), dtype=torch.bool, device=query.device
        ).tril(diagonal=0)
        attn_mask = attn_mask.masked_fill(
            torch.logical_not(attn_mask), -float("inf")
        )
        q = query.permute(0, 2, 1, 3)
        k = key.permute(0, 2, 1, 3)
        v = value.permute(0, 2, 1, 3)
        scores = torch.matmul(q, k.transpose(-2, -1)).div(3.0)
        masked_score = scores + attn_mask
        return masked_score.softmax(dim=-1).matmul(v)


def main():
    if not torch.cuda.is_available():
        raise RuntimeError("CUDA is required for this test")

    torch.manual_seed(1012)
    device = torch.device("cuda:0")
    model = Model().to(device).eval()
    inputs = tuple(
        torch.randn(4, 2, 16, 32, device=device, requires_grad=True)
        for _ in range(3)
    )

    with torch.no_grad():
        eager = model(*inputs)

    torch._dynamo.reset()
    counters.clear()
    with torch.no_grad():
        compiled = torch.compile(model, backend="inductor", fullgraph=True)(*inputs)
    torch.cuda.synchronize()
    torch.testing.assert_close(compiled, eager, atol=2e-3, rtol=2e-3)
    print("fuse_attention:", counters["inductor"]["fuse_attention"])


if __name__ == "__main__":
    main()
```

The commuted case again changes only the operand order.

```diff
- masked_score = scores + mask
+ masked_score = mask + scores
```

The same behavior occurs with `_sfdp_pattern_14`. Exchanging only the operands of the mask addition changes the result from a pattern hit to a miss.

## Root Cause

The SDPA pattern encodes `aten.add.Tensor(score, attn_mask)` with a fixed operand order. The commutative `aten.add.Tensor(attn_mask, score)` form is therefore not matched. The shared SDPA parameter check also assumes that `add.args[1]` is the attention mask.

Several other SDPA patterns—including 5, 6, 16, 19, 21, 22, 24, 26, and 29—appear to encode the same fixed operand order and may therefore be affected. These patterns were not independently tested here.

Reproduced on PyTorch main commit `f07882e`.

### Versions

PyTorch main commit f07882e

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @drisspg @liangel-02 @howardzhang-cv

### PR DESCRIPTION
**Pull request by N0AHZACH**
Fixes #195782

## Summary

The masked SDPA fusion patterns encode the attention-mask addition as `scores + attn_mask`. Because `add` is commutative, the equivalent `attn_mask + scores` form was left unfused. This PR makes the fusion match both operand orders for all masked SDPA patterns (5, 6, 14, 16, 19, 21, 22, 24, 25, 26, 29).

### `torch/_inductor/pattern_matcher.py`
The pattern matcher now tries both operand orders when matching a pattern against a commutative binary tensor op (`aten.add.Tensor`, `aten.mul.Tensor`) with two args. Only the successfully-matched operand order commits state to the match context, so a failed attempt in one order does not poison a retry in the other.

### `torch/_inductor/fx_passes/fuse_attention.py`
`_sfdp_params_check` previously assumed the attention mask was the second operand (`add.args[1]`). It now uses `_get_attn_mask_node`, which identifies the matmul-derived "scores" operand (following reshape/permute chains) regardless of position and returns the other operand as the mask. This central check is shared by all masked SDPA patterns, so the commuted form is handled everywhere.

### `test/inductor/test_fused_attention.py`
Adds `_test_sdpa_rewriter_25_commuted`, a regression test for pattern 25 using `attn_mask + scores` instead of `scores + attn_mask`, registered for CPU and GPU.

## Repro
Both `scores + mask` and `mask + scores` now report `fuse_attention: 1` and produce the fused SDPA kernel.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Review by sylvesterkaczmarek — COMMENTED** [review]
Hi, `aten.add.Tensor` is only commutative when `alpha == 1`. Swapping the operands unconditionally can make `a + alpha*b` match `b + alpha*a` for non-default alpha, which changes semantics. Could the swapped matcher be gated on the default alpha (or otherwise preserve which operand alpha scales)?

**Comment by N0AHZACH** [comment]
> Hi, `aten.add.Tensor` is only commutative when `alpha == 1`. Swapping the operands unconditionally can make `a + alpha*b` match `b + alpha*a` for non-default alpha, which changes semantics. Could the swapped matcher be gated on the default alpha (or otherwise preserve which operand alpha scales)?

Gotcha - that went out of my mind for some odd reason but yeah I did push the change thank for letting me know @sylvesterkaczmarek

**Comment by N0AHZACH** [comment]
@jansel please re-trigger the CI

**Comment by N0AHZACH** [comment]
@janel can we merge now pls

**Comment by N0AHZACH** [comment]
@jansel comeeeee onnnnn we gotto merge this.... all tests passed finally.

**Comment by N0AHZACH** [comment]
@jansel @sylvesterkaczmarek @drisspg can we please merge???? its been days

**Comment by N0AHZACH** [comment]
@jansel @drisspg @sylvesterkaczmarek - guys its been a week can someone please start the merge?

**Comment by N0AHZACH** [comment]
@drisspg @jansel @sylvesterkaczmarek @ezhang887 can we start the merge please

**Comment by N0AHZACH** [comment]
@drisspg @jansel @ezhang887 can we please start the merge - The CI tests have passed successfully

**Review by eellison — CHANGES_REQUESTED** [review]
I'm not sure how many more patterns we should add. Is this a use case that you have ? We also still need to land https://github.com/pytorch/pytorch/pull/195383 first cc @jansel
